atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/sweep/asts/bucketingthings/DefaultBucketAssignerTest.java

```diff
@@ -0,0 +1,474 @@
+        setPunchTable(1, 2);
+        bucketAssigner.run();
+
+        assertBucketState(BucketState.INCOMPLETE);
+        verify(bucketWriter).writeToAllBuckets(1, Optional.empty(), TimestampRange.of(1, -1));
+        stateTable.assertThatStateMachineTransitionedThroughTo(
+                BucketStateAndIdentifier.of(1, BucketAssignerState.waitingUntilCloseable(1)));
+    }
+
+    @Test
+    public void openingStateUsesOriginalStartTimeEvenIfPunchTableHasUpdated() {
+        stateTable.setInitialState(1, BucketAssignerState.opening(1));
+        setPunchTable(2, 3);
+        bucketAssigner.run();
+
+        assertBucketState(BucketState.INCOMPLETE);
+        // Should be 1 for start timestamp, not 2!
+        verify(bucketWriter).writeToAllBuckets(1, Optional.empty(), TimestampRange.of(1, -1));
+        stateTable.assertThatStateMachineTransitionedThroughTo(
+                BucketStateAndIdentifier.of(1, BucketAssignerState.waitingUntilCloseable(1)));
+    }
+
+    @Test
+    public void openingStateBubblesUpBucketWriteExceptionWithNoTransition() {
+        stateTable.setInitialState(1, BucketAssignerState.opening(1));
+        setPunchTable(1, 2);
+
+        when(bucketWriter.writeToAllBuckets(1, Optional.empty(), TimestampRange.of(1, -1)))
+                .thenThrow(RUNTIME_EXCEPTION);
+
+        assertThatThrownBy(bucketAssigner::run).isEqualTo(RUNTIME_EXCEPTION);
+        stateTable.assertThatStateMachineTransitionedThroughTo(); // No transition
+    }
+
+    @Test
+    public void waitingForCloseDoesNoTransitionIfBucketCannotBeClosed() {
+        stateTable.setInitialState(1, BucketAssignerState.waitingUntilCloseable(1));
+        setPunchTable(1, 2);
+        bucketAssigner.run();
+
+        assertBucketState(BucketState.INCOMPLETE);
+        verifyNoInteractions(bucketWriter);
+        stateTable.assertThatStateMachineTransitionedThroughTo(); // No transition
+    }
+
+    @Test
+    public void waitingForCloseTransitionsThroughClosingFromOpenToStartIfBucketCanBeClosed() {
+        stateTable.setInitialState(1, BucketAssignerState.waitingUntilCloseable(1));
+        setPunchTable(1, 10);
+        setPunchTableForStartWallClockTime(clock.instant().plus(Duration.ofMinutes(10)), 10, 15);
+        clock.advance(Duration.ofMinutes(10));
+        bucketAssigner.run();
+
+        assertBucketState(BucketState.CLOSED);
+        verify(bucketWriter).writeToAllBuckets(1, Optional.of(TimestampRange.of(1, -1)), TimestampRange.of(1, 10));
+        stateTable.assertThatStateMachineTransitionedThroughTo(
+                BucketStateAndIdentifier.of(1, BucketAssignerState.closingFromOpen(1, 10)),
+                // 10 was chosen as the end timestamp to show that we correctly use that as the start, and not something
+                // silly like start + 1.
+                BucketStateAndIdentifier.of(2, BucketAssignerState.start(10)),
+                // We successfully closed a bucket, so we'll run another iteration eagerly.
```

Comment:
Testing just the bucket state visitor would not require the following state checks, so might make the test a little simpler. But, the bucket assigner is pretty tightly coupled with the state visitor.

Reply:
This is maybe a little messy, but fine IMO